# <center>Vinho Verde Wine Quality Classification
* ## Stage 1: Data Preprocessing
* ### Importing Essential Libraries

In [1]:
import pandas as pd
import numpy as np

* ### Importing the Dataset

In [2]:
vinho_verde_red_url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-red.csv'
vinho_verde_white_url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-white.csv'
vinho_verde_red = pd.read_csv(vinho_verde_red_url, delimiter=';')
vinho_verde_white = pd.read_csv(vinho_verde_white_url, delimiter=';')

# standardize column names for ease of use
vinho_verde_white.columns = [c.lower().replace(' ', '_') for c in vinho_verde_white.columns]
vinho_verde_red.columns = [c.lower().replace(' ', '_') for c in vinho_verde_red.columns]

* ### Taking a peek on our Dataset

In [80]:
vinho_verde_red.head(15)

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0             7.4             0.700         0.00             1.9      0.076   
1             7.8             0.880         0.00             2.6      0.098   
2             7.8             0.760         0.04             2.3      0.092   
3            11.2             0.280         0.56             1.9      0.075   
4             7.4             0.700         0.00             1.9      0.076   
5             7.4             0.660         0.00             1.8      0.075   
6             7.9             0.600         0.06             1.6      0.069   
7             7.3             0.650         0.00             1.2      0.065   
8             7.8             0.580         0.02             2.0      0.073   
9             7.5             0.500         0.36             6.1      0.071   
10            6.7             0.580         0.08             1.8      0.097   
11            7.5             0.500         0.36             6.1      0.071   
12            5.6             0.615         0.00             1.6      0.089   
13            7.8             0.610         0.29             1.6      0.114   
14            8.9             0.620         0.18             3.8      0.176   

    free_sulfur_dioxide  total_sulfur_dioxide  density    ph  sulphates  \
0                  11.0                  34.0   0.9978  3.51       0.56   
1                  25.0                  67.0   0.9968  3.20       0.68   
2                  15.0                  54.0   0.9970  3.26       0.65   
3                  17.0                  60.0   0.9980  3.16       0.58   
4                  11.0                  34.0   0.9978  3.51       0.56   
5                  13.0                  40.0   0.9978  3.51       0.56   
6                  15.0                  59.0   0.9964  3.30       0.46   
7                  15.0                  21.0   0.9946  3.39       0.47   
8                   9.0                  18.0   0.9968  3.36       0.57   
9                  17.0                 102.0   0.9978  3.35       0.80   
10                 15.0                  65.0   0.9959  3.28       0.54   
11                 17.0                 102.0   0.9978  3.35       0.80   
12                 16.0                  59.0   0.9943  3.58       0.52   
13                  9.0                  29.0   0.9974  3.26       1.56   
14                 52.0                 145.0   0.9986  3.16       0.88   

    alcohol  quality  
0       9.4        5  
1       9.8        5  
2       9.8        5  
3       9.8        6  
4       9.4        5  
5       9.4        5  
6       9.4        5  
7      10.0        7  
8       9.5        7  
9      10.5        5  
10      9.2        5  
11     10.5        5  
12      9.9        5  
13      9.1        5  
14      9.2        5

In [81]:
vinho_verde_white.head(15)

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0             7.0              0.27         0.36           20.70      0.045   
1             6.3              0.30         0.34            1.60      0.049   
2             8.1              0.28         0.40            6.90      0.050   
3             7.2              0.23         0.32            8.50      0.058   
4             7.2              0.23         0.32            8.50      0.058   
5             8.1              0.28         0.40            6.90      0.050   
6             6.2              0.32         0.16            7.00      0.045   
7             7.0              0.27         0.36           20.70      0.045   
8             6.3              0.30         0.34            1.60      0.049   
9             8.1              0.22         0.43            1.50      0.044   
10            8.1              0.27         0.41            1.45      0.033   
11            8.6              0.23         0.40            4.20      0.035   
12            7.9              0.18         0.37            1.20      0.040   
13            6.6              0.16         0.40            1.50      0.044   
14            8.3              0.42         0.62           19.25      0.040   

    free_sulfur_dioxide  total_sulfur_dioxide  density    ph  sulphates  \
0                  45.0                 170.0   1.0010  3.00       0.45   
1                  14.0                 132.0   0.9940  3.30       0.49   
2                  30.0                  97.0   0.9951  3.26       0.44   
3                  47.0                 186.0   0.9956  3.19       0.40   
4                  47.0                 186.0   0.9956  3.19       0.40   
5                  30.0                  97.0   0.9951  3.26       0.44   
6                  30.0                 136.0   0.9949  3.18       0.47   
7                  45.0                 170.0   1.0010  3.00       0.45   
8                  14.0                 132.0   0.9940  3.30       0.49   
9                  28.0                 129.0   0.9938  3.22       0.45   
10                 11.0                  63.0   0.9908  2.99       0.56   
11                 17.0                 109.0   0.9947  3.14       0.53   
12                 16.0                  75.0   0.9920  3.18       0.63   
13                 48.0                 143.0   0.9912  3.54       0.52   
14                 41.0                 172.0   1.0002  2.98       0.67   

    alcohol  quality  
0       8.8        6  
1       9.5        6  
2      10.1        6  
3       9.9        6  
4       9.9        6  
5      10.1        6  
6       9.6        6  
7       8.8        6  
8       9.5        6  
9      11.0        6  
10     12.0        5  
11      9.7        5  
12     10.8        5  
13     12.4        7  
14      9.7        5

* ### What is the type of each feature?

In [5]:
vinho_verde_red.dtypes

fixed_acidity           float64
volatile_acidity        float64
citric_acid             float64
residual_sugar          float64
chlorides               float64
free_sulfur_dioxide     float64
total_sulfur_dioxide    float64
density                 float64
ph                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object

In [6]:
vinho_verde_white.dtypes

fixed_acidity           float64
volatile_acidity        float64
citric_acid             float64
residual_sugar          float64
chlorides               float64
free_sulfur_dioxide     float64
total_sulfur_dioxide    float64
density                 float64
ph                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object

* ### Getting a Basic Statistical Description of each Dataset

In [7]:
vinho_verde_white.describe()

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  \
count    4898.000000       4898.000000  4898.000000     4898.000000   
mean        6.854788          0.278241     0.334192        6.391415   
std         0.843868          0.100795     0.121020        5.072058   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.300000          0.210000     0.270000        1.700000   
50%         6.800000          0.260000     0.320000        5.200000   
75%         7.300000          0.320000     0.390000        9.900000   
max        14.200000          1.100000     1.660000       65.800000   

         chlorides  free_sulfur_dioxide  total_sulfur_dioxide      density  \
count  4898.000000          4898.000000           4898.000000  4898.000000   
mean      0.045772            35.308085            138.360657     0.994027   
std       0.021848            17.007137             42.498065     0.002991   
min       0.009000             2.000000              9.000000     0.987110   
25%       0.036000            23.000000            108.000000     0.991723   
50%       0.043000            34.000000            134.000000     0.993740   
75%       0.050000            46.000000            167.000000     0.996100   
max       0.346000           289.000000            440.000000     1.038980   

                ph    sulphates      alcohol      quality  
count  4898.000000  4898.000000  4898.000000  4898.000000  
mean      3.188267     0.489847    10.514267     5.877909  
std       0.151001     0.114126     1.230621     0.885639  
min       2.720000     0.220000     8.000000     3.000000  
25%       3.090000     0.410000     9.500000     5.000000  
50%       3.180000     0.470000    10.400000     6.000000  
75%       3.280000     0.550000    11.400000     6.000000  
max       3.820000     1.080000    14.200000     9.000000

In [8]:
vinho_verde_red.describe()

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free_sulfur_dioxide  total_sulfur_dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             22.000000     0.995600   
50%       0.079000            14.000000             38.000000     0.996750   
75%       0.090000            21.000000             62.000000     0.997835   
max       0.611000            72.000000            289.000000     1.003690   

                ph    sulphates      alcohol      quality  
count  1599.000000  1599.000000  1599.000000  1599.000000  
mean      3.311113     0.658149    10.422983     5.636023  
std       0.154386     0.169507     1.065668     0.807569  
min       2.740000     0.330000     8.400000     3.000000  
25%       3.210000     0.550000     9.500000     5.000000  
50%       3.310000     0.620000    10.200000     6.000000  
75%       3.400000     0.730000    11.100000     6.000000  
max       4.010000     2.000000    14.900000     8.000000

* ### Functions
  * **find_outlier_indices:** Given a DataFrame return the outlier indices in list of tuples in (feature, row_indices) format, *provided that the feature values are greater than Q75 + 1.5 IQR or less that Q25 - 1.5 IQR*
  * **drop_outliers:** Given a DataFrame and the Outlier list of tuples in (feature, row_indices) format, return the DataFrame minus the included in the list row indices

In [9]:
def find_outlier_indices(df):
    quantiles = df.quantile([0.25, 0.75])
    feature_outlier_indices = []
    for feature in quantiles:
        q25, q75 = quantiles[feature].values
        iqr = q75 - q25
        
        outlier_indices = df.index[(df[feature] < q25 - (1.5 * iqr)) | (df[feature] > q75 + (1.5 * iqr))].tolist()
        feature_outlier_indices.append((feature, outlier_indices))
    return feature_outlier_indices

def drop_outliers(df, outliers):
    indices = set([])
    for col, col_indices in outliers:
        indices = indices.union(set(col_indices))
    return df.drop(list(indices))

* ### 1.1. Detecting NaN/Empty Values

In [18]:
for feature in vinho_verde_white:
    print ('Feature: {0} === {1} NaN Values, {2} Missing Values.'.format(feature, np.sum(vinho_verde_white[feature] == np.nan),  np.sum(vinho_verde_white[feature].isnull())))

Feature: fixed_acidity === 0 NaN Values, 0 Missing Values.
Feature: volatile_acidity === 0 NaN Values, 0 Missing Values.
Feature: citric_acid === 0 NaN Values, 0 Missing Values.
Feature: residual_sugar === 0 NaN Values, 0 Missing Values.
Feature: chlorides === 0 NaN Values, 0 Missing Values.
Feature: free_sulfur_dioxide === 0 NaN Values, 0 Missing Values.
Feature: total_sulfur_dioxide === 0 NaN Values, 0 Missing Values.
Feature: density === 0 NaN Values, 0 Missing Values.
Feature: ph === 0 NaN Values, 0 Missing Values.
Feature: sulphates === 0 NaN Values, 0 Missing Values.
Feature: alcohol === 0 NaN Values, 0 Missing Values.
Feature: quality === 0 NaN Values, 0 Missing Values.


In [11]:
for feature in vinho_verde_red:
    print ('Feature: {0} === {1} NaN Values, {2} Misisng Values.'.format(feature, np.sum(vinho_verde_red[feature] == np.nan),  np.sum(vinho_verde_red[feature].isnull())))

Feature: fixed_acidity === 0 NaN Values, 0 Misisng Values.
Feature: volatile_acidity === 0 NaN Values, 0 Misisng Values.
Feature: citric_acid === 0 NaN Values, 0 Misisng Values.
Feature: residual_sugar === 0 NaN Values, 0 Misisng Values.
Feature: chlorides === 0 NaN Values, 0 Misisng Values.
Feature: free_sulfur_dioxide === 0 NaN Values, 0 Misisng Values.
Feature: total_sulfur_dioxide === 0 NaN Values, 0 Misisng Values.
Feature: density === 0 NaN Values, 0 Misisng Values.
Feature: ph === 0 NaN Values, 0 Misisng Values.
Feature: sulphates === 0 NaN Values, 0 Misisng Values.
Feature: alcohol === 0 NaN Values, 0 Misisng Values.
Feature: quality === 0 NaN Values, 0 Misisng Values.


<center>_In that occation we didn't had any NaN or Missing Values, so we move forward to detect (and remove) the Outliers that (may) exist in the dataset_

* ### 1.2. Searching (and Removing) Outliers

In [12]:
feature_outlier_indices = find_outlier_indices(vinho_verde_white)
for feature, feature_indices in feature_outlier_indices:
    print ('Feature: {0} === Outlier Values Count: {1}'.format(feature, len(feature_indices)))

Feature: fixed_acidity === Outlier Values Count: 119
Feature: volatile_acidity === Outlier Values Count: 186
Feature: citric_acid === Outlier Values Count: 270
Feature: residual_sugar === Outlier Values Count: 7
Feature: chlorides === Outlier Values Count: 212
Feature: free_sulfur_dioxide === Outlier Values Count: 50
Feature: total_sulfur_dioxide === Outlier Values Count: 19
Feature: density === Outlier Values Count: 5
Feature: ph === Outlier Values Count: 75
Feature: sulphates === Outlier Values Count: 124
Feature: alcohol === Outlier Values Count: 0
Feature: quality === Outlier Values Count: 200


In [13]:
feature_outlier_indices = find_outlier_indices(vinho_verde_red)
for feature, feature_indices in feature_outlier_indices:
    print ('Feature: {0} === Outlier Values Count: {1}'.format(feature, len(feature_indices)))

Feature: fixed_acidity === Outlier Values Count: 49
Feature: volatile_acidity === Outlier Values Count: 19
Feature: citric_acid === Outlier Values Count: 1
Feature: residual_sugar === Outlier Values Count: 155
Feature: chlorides === Outlier Values Count: 112
Feature: free_sulfur_dioxide === Outlier Values Count: 30
Feature: total_sulfur_dioxide === Outlier Values Count: 55
Feature: density === Outlier Values Count: 45
Feature: ph === Outlier Values Count: 35
Feature: sulphates === Outlier Values Count: 59
Feature: alcohol === Outlier Values Count: 13
Feature: quality === Outlier Values Count: 28


#### *Before Outlier Removal*

In [14]:
vinho_verde_white.shape

(4898, 12)

In [15]:
vinho_verde_red.shape

(1599, 12)

#### *After Outlier Removal*

In [16]:
vinho_verde_white_prep = drop_outliers(vinho_verde_white, feature_outlier_indices)
vinho_verde_white_prep.shape

(4478, 12)

In [17]:
vinho_verde_red_prep = drop_outliers(vinho_verde_red, feature_outlier_indices)
vinho_verde_red_prep.shape

(1179, 12)

* ### 1.3. Correlation Analysis 

In [56]:
vv_white_corr = vinho_verde_white.corr().abs()
vv_white_prep_corr = vinho_verde_white_prep.corr().abs() # Outlier Removed
vv_red_corr = vinho_verde_red.corr().abs()
vv_red_prep_corr = vinho_verde_red_prep.corr().abs() # Outlier Removed

* #### Correlation Analysis on *(Outlier Removed)* Red Wine Samples

In [76]:
vv_red_features = vinho_verde_red.shape[1]
vv_red_prep_features = vinho_verde_red_prep.shape[1]

s = vv_red_corr.unstack().sort_values(ascending=False, kind="mergesort")
print ('Correlation Analysis: Red Wine\n\n', s[vv_red_features:], '\n\n')

s = vv_red_prep_corr.unstack().sort_values(ascending=False, kind="mergesort")
print ('Correlation Analysis: Red Wine - Outlier Free\n\n', s[vv_red_prep_features:])

Correlation Analysis: Red Wine

 ph                    fixed_acidity           0.682978
fixed_acidity         ph                      0.682978
citric_acid           fixed_acidity           0.671703
fixed_acidity         citric_acid             0.671703
density               fixed_acidity           0.668047
fixed_acidity         density                 0.668047
total_sulfur_dioxide  free_sulfur_dioxide     0.667666
free_sulfur_dioxide   total_sulfur_dioxide    0.667666
citric_acid           volatile_acidity        0.552496
volatile_acidity      citric_acid             0.552496
ph                    citric_acid             0.541904
citric_acid           ph                      0.541904
alcohol               density                 0.496180
density               alcohol                 0.496180
quality               alcohol                 0.476166
alcohol               quality                 0.476166
quality               volatile_acidity        0.390558
volatile_acidity      quality   

##### Most Correlated Features
* ph                    fixed_acidity           **0.682978**, Outlier Removed: **0.684378**
* citric_acid           fixed_acidity           **0.671703**, Outlier Removed: **0.659397**
* density               fixed_acidity           **0.668047**, Outlier Removed: **0.627194**
* total_sulfur_dioxide  free_sulfur_dioxide     **0.667667**, Outlier Removed: **0.619675**
* citric_acid           volatile_acidity        **0.552496**, Outlier Removed: **0.610125**
* ph                    citric_acid             **0.541904**, Outlier Removed: **0.538901**

* #### Correlation Analysis on *(Outlier Removed)* White Wine Samples

In [75]:
vv_white_features = vinho_verde_white.shape[1]
vv_white_prep_features = vinho_verde_white_prep.shape[1]

s = vv_white_corr.unstack().sort_values(ascending=False, kind="mergesort")
print ('Correlation Analysis: White Wine\n\n', s[vv_white_features:], '\n\n')

s = vv_white_prep_corr.unstack().sort_values(ascending=False, kind="mergesort")
print ('Correlation Analysis: White Wine - Outlier Free\n\n', s[vv_white_prep_features:], '\n\n')

Correlation Analysis: White Wine

 density               residual_sugar          0.838966
residual_sugar        density                 0.838966
alcohol               density                 0.780138
density               alcohol                 0.780138
total_sulfur_dioxide  free_sulfur_dioxide     0.615501
free_sulfur_dioxide   total_sulfur_dioxide    0.615501
density               total_sulfur_dioxide    0.529881
total_sulfur_dioxide  density                 0.529881
alcohol               residual_sugar          0.450631
residual_sugar        alcohol                 0.450631
alcohol               total_sulfur_dioxide    0.448892
total_sulfur_dioxide  alcohol                 0.448892
quality               alcohol                 0.435575
alcohol               quality                 0.435575
ph                    fixed_acidity           0.425858
fixed_acidity         ph                      0.425858
total_sulfur_dioxide  residual_sugar          0.401439
residual_sugar        total_su

##### Most Correlated Features
* density               residual_sugar          **0.838966**, Outlier Removed: **0.838270**
* alcohol               density                 **0.780138**, Outlier Removed: **0.781072**
* total_sulfur_dioxide  free_sulfur_dioxide     **0.615501**, Outlier Removed: **0.613695**
* density               total_sulfur_dioxide    **0.529881**, Outlier Removed: **0.530652**

In [88]:
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib import style;  style.use('ggplot')

# TODO

* ### 1.4. Feature Normalization

* ### 1.5. *(If Applicable)* PCA Dimensionality Reduction

* ### 1.6. Dataset Sampling (and use Pandas describe() function to compare with original's stats)

* ### 1.6. T-SNE Dataset 2-D Plot (Before/After Sampling)

* ### 2. Save the Dataset for use in future Steps. The Dataset should be:
  * ##### cleaned *(free from NaN/Null values)*
  * ##### normalized
  * ##### dimension reduced *(if applicable)*
  * ##### non-outlier removed *(we'll see about that)*
  * ##### non-sampled *(the dataset is already small enough)*